### PR TITLE
Implement STAC and OGC API compatibility improvements

### LINKED ISSUE #904: STAC and OGC API compatibility improvements
Other OGC Testbed-19 participants are now actively testing the xcube OGC Data Cube API implementation, which comprises the STAC and OGC API - Coverages server plug-ins. Various minor bugs and compatibility issues are being reported. This issue, and its associated PR, are collecting a batch of fixes and improvements which are too minor to deserve individual PRs:

- Expand the STAC catalogue coverage links to provide links for multiple formats.
- For TIFF and PNG OGC coverages, apply the squeeze method to remove any singleton dimensions.
- Add support for the scale-factor parameter to the main coverage endpoint.
- Improve handling of bbox parameters: they are now applied to the axes in the order that they appear in the dataset, as required in the specification. Inverted y-axes are also handled.
- Handle half-open datetime intervals.
- More parameter checking: various invalid parameters will now result in a 400 response rather than a server-side error.
- Produce a 501 response for parameters which are not yet implemented.
- Add media types to coverage links.
- Return a 400 Bad Request response for attempts at subsetting on non-existent axes.
- Handle quotation marks in coverages subset parameter, if present.
- Use crs_wkt when determining CRS, if present and needed.
- Give a 400 Bad Request response on requests for non-existent properties (i.e. variables).
- Change default subsetting and bbox CRS from EPSG:4326 to OGC:CRS84.
- Implement reprojection for bbox.
- Run datetime parameter values through `ensure_time_index_compatible` function to fix timezone-aware / -naive incompatibilities
- If writing an image from a Dataset with a single variable, write the relevant DataArray rather than the whole Dataset (avoids errors due to additional "bnds" dimensions).
- Return the appropriate HTTP responses for requests which result in no data or too much data. 
- Add `crs` and `crs_storage` properties to STAC data.
- Reimplement subsetting to conform more closely to the standard and to separate request parsing more cleanly from data operations.
- Set the Content-Bbox and Content-Crs headers when returning a coverage.
- Support safe CURIE syntax for CRS specification.

### PR DESCRIPTION
**Pull request by pont-us**
Closes #904 .

- Expand the STAC catalogue coverage links to provide links for multiple formats.
- For TIFF and PNG OGC coverages, apply the squeeze method to remove any singleton dimensions.
- Add support for the scale-factor parameter to the main coverage endpoint.
- Improve handling of bbox parameters: they are now applied to the axes in the order that they appear in the dataset, as required in the specification. Inverted y-axes are also handled.
- Handle half-open datetime intervals.
- More parameter checking: various invalid parameters will now result in a 400 response rather than a server-side error.
- Produce a 501 response for parameters which are not yet implemented.
- Add media types to coverage links.
- Return a 400 Bad Request response for attempts at subsetting on non-existent axes.
- Handle quotation marks in coverages subset parameter, if present.
- Use crs_wkt when determining CRS, if present and needed.
- Give a 400 Bad Request response on requests for non-existent properties (i.e. variables).
- Change default subsetting and bbox CRS from EPSG:4326 to OGC:CRS84.
- Implement reprojection for bbox.
- Run datetime parameter values through `ensure_time_index_compatible` function to fix timezone-aware / -naive incompatibilities
- If writing an image from a Dataset with a single variable, write the relevant DataArray rather than the whole Dataset (avoids errors due to additional "bnds" dimensions).
- Return the appropriate HTTP responses for requests which result in no data or too much data. 
- Add `crs` and `crs_storage` properties to STAC data.
- Reimplement subsetting to conform more closely to the standard and to separate request parsing more cleanly from data operations.
- Set the Content-Bbox and Content-Crs headers when returning a coverage.
- Support safe CURIE syntax for CRS specification.

Checklist:

* [x] Add unit tests and/or doctests in docstrings
* [x] Add docstrings and API docs for any new/modified user-facing classes and functions
* [x] New/modified features documented in `docs/source/*`
* [x] Changes documented in `CHANGES.md`
* [x] GitHub CI passes
* [x] AppVeyor CI passes
* [x] Test coverage remains or increases (target 100%)

### CODE REVIEW

File: xcube/webapi/ows/coverages/controllers.py [hunk 1]
```diff
@@ -65,7 +67,7 @@ def get_coverage_data(
     collection_id: str,
     query: Mapping[str, Sequence[str]],
     content_type: str,
-) -> Optional[bytes]:
+) -> tuple[Optional[bytes], list[float], pyproj.CRS]:
```
Comment (TonioF) [inline xcube/webapi/ows/coverages/controllers.py:70]: Please adjust doc

File: xcube/webapi/ows/coverages/controllers.py [hunk 2]
```diff
@@ -144,34 +190,270 @@ def get_coverage_data(
...
-            # TODO: don't flip longitude
-            if (low < high) != (ds[axis][0] < ds[axis][-1]):
-                low, high = high, low
-            if axis != 'time':
-                # TODO: catch parse exception and pass through as string
+def _get_crs_axis_by_name(crs: pyproj.CRS, name: str):
+    for axis in crs.axis_info:
+        if axis.abbrev == name or axis.name == name:
+            return axis
+    return None
```
Comment (TonioF) [inline xcube/webapi/ows/coverages/controllers.py]: It seems this function is not used anywhere.

File: xcube/webapi/ows/coverages/controllers.py [hunk 3]
```diff
@@ -144,34 +190,270 @@ def get_coverage_data(
...
+    for axis, value in subset_spec.items():
+        if isinstance(value, str):  # single value
+            if axis == 'time':
+                specifiers[axis] = ensure_time_index_compatible(ds, value)
+            else:
+                try:
+                    # Parse to float if possible
+                    specifiers[axis] = float(value)
+                except ValueError:
+                    specifiers[axis] = value
```
Comment (TonioF) [inline xcube/webapi/ows/coverages/controllers.py:268]: Sure it is no problem if it is not a float?

File: xcube/webapi/ows/coverages/controllers.py [hunk 4]
```diff
@@ -144,34 +190,270 @@ def get_coverage_data(
...
+) -> list[float]:
+    if source_crs == dest_crs:
+        return bbox
+    else:
+        transformer = pyproj.Transformer.from_crs(
+            source_crs, dest_crs, always_xy=True
+        )
+        bbox_ = bbox.copy()
+        _ensure_bbox_y_ascending(bbox_)
+        return list(transformer.transform_bounds(*bbox_))
```
Comment (TonioF) [inline xcube/webapi/ows/coverages/controllers.py:382]: You can remove line 379 and move everything up one indent :)

File: xcube/webapi/ows/coverages/controllers.py [hunk 5]
```diff
@@ -191,9 +473,17 @@ def dataset_to_image(
...
                 ds[var] = ds[var].astype(np.uint16, casting='unsafe')
 
+    ds = ds.squeeze()
+
     with tempfile.TemporaryDirectory() as tempdir:
         path = os.path.join(tempdir, 'out.' + image_format)
-        ds.rio.to_raster(path)
+        ds = ds.drop_vars(
+            names=['crs', 'spatial_ref'], errors='ignore'
+        ).squeeze()
```
Comment (TonioF) [inline xcube/webapi/ows/coverages/controllers.py:464]: Why is this necessary?

File: xcube/webapi/ows/stac/context.py [hunk 6]
```diff
@@ -33,6 +33,15 @@ def __init__(self, server_ctx: Context):
         self._datasets_ctx = server_ctx.get_api_ctx('datasets',
                                                     cls=DatasetsContext)
 
+        # TODO: decide what else to include in this list.
+        #  Listing all 11396 CRSs currently known to pyproj seems
+        #  impractical. Make it a configuration option, perhaps?
+        self._available_crss = ['OGC:CRS84', 'EPSG:4326']
```
Comment (TonioF) [inline xcube/webapi/ows/stac/context.py:39]: Is not even trying to list them and instead just validating what is handed in an option?

File: test/webapi/ows/stac/test_controllers.py [hunk 7]
```diff
@@ -38,11 +39,12 @@
 from xcube.webapi.ows.stac.config import DEFAULT_COLLECTION_DESCRIPTION
 from xcube.webapi.ows.stac.config import DEFAULT_COLLECTION_ID
 from xcube.webapi.ows.stac.config import DEFAULT_COLLECTION_TITLE
+from xcube.webapi.ows.stac.context import StacContext
 from xcube.webapi.ows.stac.controllers import (
     STAC_VERSION,
     get_collection_queryables,
     get_datacube_dimensions,
-    get_single_collection_items,
+    get_single_collection_items, crs_to_uri_or_wkt,
```
Comment (TonioF) [inline test/webapi/ows/stac/test_controllers.py:47]:
```suggestion
    get_single_collection_items, 
    crs_to_uri_or_wkt,
```

File: xcube/webapi/ows/stac/controllers.py [hunk 8]
```diff
@@ -482,47 +495,76 @@ def _get_single_dataset_collection(
...
         'type': 'Collection',
+        'storageCRS': storage_crs,
+        'crs': available_crss
     }
-    result.update(_get_cube_properties(ctx, dataset_id))
+    result.update(_get_cube_properties(ds_ctx, dataset_id))
     return result
 
 
+def crs_to_uri_or_wkt(crs: pyproj.CRS) -> str:
```
Comment (TonioF) [inline xcube/webapi/ows/stac/controllers.py:545]: I'm not a huge fan of functions where you don't know what you'll get. I acknowledge for its purpose it's fine but still ...

File: xcube/webapi/ows/stac/controllers.py [hunk 9]
```diff
@@ -482,47 +495,76 @@ def _get_single_dataset_collection(
             {
                 'rel': 'items',
                 'href': f'{base_url}{PATH_PREFIX}/collections/'
-                        f'{dataset_id}/items',
+                f'{dataset_id}/items',
```
Comment (TonioF) [inline xcube/webapi/ows/stac/controllers.py:498]: I preferred the previous format (same applies for entries below).

**Review by TonioF — CHANGES_REQUESTED** [review]
I did not verify that your pr solves every single bullet listed in the issue - it looks good in general, though. Still, I could dig up some things you might want to look at.

File: CHANGES.md [hunk 10]
```diff
@@ -1,5 +1,25 @@
...
+  * Robuster, more standard-compliant parameter parsing and checking.
+  * More informative responses for incorrect or unsupported parameters.
+  * Omit unnecessary dimensions in TIFF and PNG coverages.
+  * Use crs_wkt when determining CRS, if present and needed.
+  * Change default subsetting and bbox CRS from EPSG:4326 to OGC:CRS84.
+  * Implement reprojection for bbox.
+  * Ensure datetime parameters match dataset’s timezone awareness.
+  * Reimplement subsetting (better standards conformance, cleaner code).
+  * Set Content-Bbox and Content-Crs headers in the HTTP response.
+  * Support safe CURIE syntax for CRS specification.
```
[the remark this answers is not in the document]
Comment (pont-us) [reply CHANGES.md]: Happy to oblige, since I have no personal preference. (Incidentally, though, I maintain that they are full sentences, in the imperative mood. I don't think we have a defined best practice in xcube for this, so I tend to go by the [git patch submission guidelines](https://git.kernel.org/pub/scm/git/git.git/tree/Documentation/SubmittingPatches?h=v2.36.1#n181): ‘Describe your changes in imperative mood... as if you are giving orders to the codebase to change its behavior.’)

File: xcube/webapi/ows/stac/context.py (hunk 6 above)
Comment (pont-us) [reply xcube/webapi/ows/stac/context.py:39]: That is effectively the current behaviour. Requests aren't validated against this list, and the server will attempt to parse and use whatever CRS specifier the client provides.

File: xcube/webapi/ows/stac/controllers.py (hunk 9 above)
Comment (pont-us) [reply xcube/webapi/ows/stac/controllers.py]: So do I. Sorry about all the pure formatting changes in this file. I would never deliberately add them to such a large PR, but PyCharm "helpfully" did this for me while editing and by the time I noticed, its "improvements" were so mixed up with my own code changes that I could only undo them by hand -- and I missed quite a few, such as that one. I've now fixed all (I think) the "weird string indent" reformattings as well as a few others, which should make things marginally less painful for the next reviewers.

File: xcube/webapi/ows/stac/controllers.py (hunk 8 above)
Comment (pont-us) [reply xcube/webapi/ows/stac/controllers.py:545]: I agree it's not pretty, but I can't think of a better way to do it: we want to use a standard URI if a CRS is representable as such, and fall back to a WKT representation if not. The caller at least knows that they'll get a valid string representation of the supplied CRS, and under the circumstances I think that's good enough.

File: xcube/webapi/ows/coverages/controllers.py [hunk 11]
```diff
@@ -81,58 +83,102 @@ def get_coverage_data(
...
+
+    if request.subset is not None:
+        subset_bbox, ds = _apply_subsetting(ds, request.subset, subset_crs)
+    else:
+        subset_bbox = None
+
+    if request.bbox is not None:
+        ds = _apply_bbox(ds, request.bbox, bbox_crs, always_xy=False)
+
+    # ds.rio.write_crs(get_crs_from_dataset(ds), inplace=True)
```
[the remark this answers is not in the document]
Comment (pont-us) [reply xcube/webapi/ows/coverages/controllers.py]: No -- removed.

File: xcube/webapi/ows/coverages/controllers.py (hunk 4 above)
Comment (pont-us) [reply xcube/webapi/ows/coverages/controllers.py]: A matter of taste, but OK, done :)

File: xcube/webapi/ows/coverages/controllers.py (hunk 5 above)
Comment (pont-us) [reply xcube/webapi/ows/coverages/controllers.py:467]: The conversion to TIFF/PNG doesn't work if there are extra variables or dimensions. I've added a comment mentioning this.

**Review by TonioF — APPROVED** [review]
I'm good now :) .

**Review by thomasstorm — APPROVED** [review]
Only a question/suggestion, otherwise approved. Nice work!

File: xcube/webapi/ows/coverages/controllers.py [hunk 12]
```diff
@@ -144,34 +194,249 @@ def get_coverage_data(
...
-        if high is None:
-            # TODO: parse to float if possible
-            indices[axis] = low
-        else:
-            # TODO: don't flip longitude
-            if (low < high) != (ds[axis][0] < ds[axis][-1]):
-                low, high = high, low
-            if axis != 'time':
-                # TODO: catch parse exception and pass through as string
+def _apply_subsetting(
```
[the remark this answers is not in the document]
Comment (pont-us) [reply xcube/webapi/ows/coverages/controllers.py:236]: Mmmaybe. But it's currently quite tightly coupled to the OGC API - Coverages model -- amongst other things, this means that the specifier mixes non-spatial specifiers (which refer directly to a native dataset dimension) and spatial specifiers (which define a bounding box or other area in an arbitrary CRS which may have quite a complex transformation to the nearest equivalent subset in the native dataset). We have to implement this for conformance, but I'm not convinced by either the conceptual elegance or the general usefulness of the approach (and not sure that OGC are either, since they also mandate a bbox parameter!).

I suggest that we wait for the planned unification of the WCS and OGC API - Coverages implementations, which will naturally lead to code consolidation. Whatever we consolidate there we can also consider moving into the core.

File: test/webapi/ows/coverages/test_request.py [hunk 13]
```diff
@@ -0,0 +1,134 @@
+import unittest
+import pyproj
+
+from xcube.webapi.ows.coverages.request import CoveragesRequest
+
+
+class CoveragesUtilTest(unittest.TestCase):
```
Comment (forman) [inline test/webapi/ows/coverages/test_request.py:7]:
```suggestion
class CoveragesRequestTest(unittest.TestCase):
```

File: xcube/webapi/ows/coverages/controllers.py [hunk 14]
```diff
@@ -77,62 +79,110 @@ def get_coverage_data(
...
+    if native_crs != final_crs and request.bbox is not None:
+        ds = _apply_bbox(ds, request.bbox, bbox_crs, always_xy=False)
+    if subset_bbox is not None:
+        ds = _apply_bbox(ds, subset_bbox, subset_crs, always_xy=True)
+
+    ds.rio.write_crs(final_crs, inplace=True)
+    for var in ds.data_vars.values():
+        var.attrs.pop('grid_mapping', None)
+
+    # TODO rename axes to match final CRS?
```
Comment (forman) [inline xcube/webapi/ows/coverages/controllers.py:173]: Is this really a TODO? Otherwise

```suggestion
    # check: rename axes to match final CRS?
```

File: xcube/webapi/ows/stac/context.py [hunk 15]
```diff
@@ -33,6 +33,15 @@ def __init__(self, server_ctx: Context):
         self._datasets_ctx = server_ctx.get_api_ctx('datasets',
                                                     cls=DatasetsContext)
 
+        # TODO: decide what else to include in this list.
+        #  Listing all 11396 CRSs currently known to pyproj seems
+        #  impractical. Make it a configuration option, perhaps?
```
Comment (forman) [inline xcube/webapi/ows/stac/context.py:38]: Is this really a TODO? If so, say why this is needed.

Actually I don't understand why we limit it to geographic. xcube can reproject to any CRS, see tiles generation.

Comment (pont-us) [reply xcube/webapi/ows/stac/context.py]: I've downgraded it to a "check".
